### ISSUE TITLE
Configure USB gadget device before BEAM boot

### ISSUE THREAD

**Issue by amclain** [ElixirSeattle/xebow#118]
We would like to configure the USB gadget device before the BEAM boots so that in the future we can have access to it from a separate application that can send simple HID keyboard reports, like pressing a key to get into the host's BIOS. Due to this we also need to migrate off of `usb_gadget` since we won't have access to Elixir at this stage.

**Comment by amclain** [comment]
Update:
- [`gt` (gadget tool)](https://github.com/kopasiak/gt) can configure configFS USB gadgets via a declarative syntax.
- I pushed up what that it looks like for [our config](https://github.com/ElixirSeattle/xebow/blob/refactor/usb-gadget-gt/rootfs_overlay/etc/gt/templates/default#L49), but `report_desc` for the keyboard HID device isn't recognized by `gt` and I don't think they have a way to specify a blob of binary data yet.
- I have a [`nerves_system_keybow` branch](https://github.com/ElixirSeattle/nerves_system_keybow/tree/refactor/usb-gadget-gt) up for `gt`.
- As well as a [`nerves_system_br` branch](https://github.com/amclain/nerves_system_br/tree/gt) with the `gt` package added to buildroot.
- I worked up an [alternative branch](https://github.com/ElixirSeattle/xebow/tree/refactor/usb-gadget-script/rootfs_overlay/etc/usb_gadget) that uses a script to configure the USB gadget.
- Our system is missing the `ln` command which is needed to use configFS, so I've added it to a [`nerves_system_keybow` branch](https://github.com/ElixirSeattle/nerves_system_keybow/tree/refactor/usb-gadget-script).
- I'm trying to get `gt` OR the script to run before the BEAM starts. That is defined in [erlinit.config](https://github.com/ElixirSeattle/xebow/blob/refactor/usb-gadget-script/rootfs_overlay/etc/erlinit.config#L65-L66).
- The USB device can be defined at this stage, but not enabled. I think the UDC (USB device controller) is not available at this stage for some reason. The same script works just fine in `Application.start`. Error messages at this point are spotty.
- I'm digging through the `erlinit`, `nerves_runtime`, `nerves_pack`, `vintage_net` chain to see if anything is going on with the UDC between the time it doesn't work and the time it does.

**Comment by amclain** [comment]
```
00:00:23.211 [warn]  dwc2 20980000.usb: 20980000.usb supply vusb_d not found, using dummy regulator

00:00:23.211 [info]  dwc2 20980000.usb: Linked as a consumer to regulator.0

00:00:23.212 [warn]  dwc2 20980000.usb: 20980000.usb supply vusb_a not found, using dummy regulator

00:00:23.213 [warn]  dwc2 20980000.usb: Configuration mismatch. dr_mode forced to device

00:00:23.321 [warn]  dwc2 20980000.usb: dwc2_check_params: Invalid parameter lpm=1

00:00:23.321 [warn]  dwc2 20980000.usb: dwc2_check_params: Invalid parameter lpm_clock_gating=1

00:00:23.322 [warn]  dwc2 20980000.usb: dwc2_check_params: Invalid parameter besl=1

00:00:23.323 [warn]  dwc2 20980000.usb: dwc2_check_params: Invalid parameter hird_threshold_en=1
 
00:00:23.323 [info]  dwc2 20980000.usb: EPs: 8, dedicated fifos, 4080 entries in SPRAM
```

**Comment by doughsay** [comment]
Shouldn't UDC be available after kernel boot and have nothing to do with beam/nerves booting?  This sounds more like a Linux problem than a nerves problem...

**Comment by amclain** [comment]
>This sounds more like a Linux problem than a nerves problem...

tl;dr yes. I think one of the Nerves packages is doing the initialization that we want to do sooner, so I'm trying to track that down and move the logic to before the BEAM starts.

**Comment by amclain** [comment]
Update:
I discovered `gadget-import` as part of [`libusbgx`](https://github.com/libusbgx/libusbgx), which can import declarative schemas. It also has the inverse tool called `gadget-export` which was able to export the binary descriptor for the HID keyboard device's `report_desc` field. I also built a new `nerves_system_keybow` and things seem to be working.
